feat: add verified wallet accounts to community requests

0x-r4bbit opened the pull request:
This commit extends the `CommunityRequestToJoin` with `RevealedAddresses` which represent wallet addresses and signatures provided by the sender, to proof a community owner ownership of those wallet addresses.

**Note: This only works with keystore files maanged by status-go**

At high level, the follwing happens:

1. User instructs Status to send a request to join to a community. By adding a password hash to the instruction, Status will try to unlock the users keystore and verify each wallet account.
2. For every verified wallet account, a signature is created for the following payload, using each wallet's private key

   ``` keccak256(chatkey + communityID + requestToJoinID) ``` A map of walletAddress->signature is then attached to the community request to join, which will be sent to the community owner
3. The owner node receives the request, and if the community requires users to hold tokens to become a member, it will check and verify whether the given wallet addresses are indeed owned by the sender. If any signature provided by the request cannot be recovered, the request is immediately declined by the owner.
4. The verified addresses are then added to the owner node's database such that, once the request should be accepted, the addresses can be used to check on chain whether they own the necessary funds to fulfill the community's permissions

The checking of required funds is **not** part of this commit. It will be added in a follow-up commit.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (4)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 3849f894 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3295/1/) | 2023-03-16 12:36:11 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230316-123453-3849f8-pr3295.zip) |
| :heavy_multiplication_x: | 3849f894 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/1/) | 2023-03-16 12:37:03 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/1//consoleText) |
| :heavy_check_mark: | 3849f894 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3295/1/) | 2023-03-16 12:39:05 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230316-123453-3849f8-pr3295.zip) |
| :heavy_check_mark: | 3849f894 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3295/1/) | 2023-03-16 12:39:24 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230316-123448-3849f8-pr3295.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | d0286a38 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3295/2/) | 2023-03-17 09:26:25 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230317-092501-d0286a-pr3295.zip) |
| :heavy_check_mark: | d0286a38 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3295/2/) | 2023-03-17 09:27:56 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230317-092457-d0286a-pr3295.zip) |
| :heavy_check_mark: | d0286a38 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3295/2/) | 2023-03-17 09:28:33 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230317-092457-d0286a-pr3295.aar) |
| :heavy_check_mark: | d0286a38 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/2/) | 2023-03-17 09:36:28 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | bf380b13 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3295/3/) | 2023-03-22 11:29:31 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230322-112812-bf380b-pr3295.zip) |
| :heavy_check_mark: | bf380b13 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3295/3/) | 2023-03-22 11:32:28 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230322-112812-bf380b-pr3295.aar) |
| :heavy_check_mark: | bf380b13 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3295/3/) | 2023-03-22 11:32:48 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230322-112812-bf380b-pr3295.zip) |
| :heavy_multiplication_x: | bf380b13 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/3/) | 2023-03-22 11:35:44 | ~7 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/3//consoleText) |
| :heavy_check_mark: | bf380b13 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/4/) | 2023-03-22 12:06:48 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3295/4//consoleText) |

0x-r4bbit committed:
feat: add verified wallet accounts to community requests

This commit extends the `CommunityRequestToJoin` with `RevealedAddresses` which represent wallet addresses and signatures provided by the sender, to proof a community owner ownership of those wallet addresses.

**Note: This only works with keystore files maanged by status-go**

At high level, the follwing happens:

1. User instructs Status to send a request to join to a community. By adding a password hash to the instruction, Status will try to unlock the users keystore and verify each wallet account.
2. For every verified wallet account, a signature is created for the following payload, using each wallet's private key

   ``` keccak256(chatkey + communityID + requestToJoinID) ``` A map of walletAddress->signature is then attached to the community request to join, which will be sent to the community owner
3. The owner node receives the request, and if the community requires users to hold tokens to become a member, it will check and verify whether the given wallet addresses are indeed owned by the sender. If any signature provided by the request cannot be recovered, the request is immediately declined by the owner.
4. The verified addresses are then added to the owner node's database such that, once the request should be accepted, the addresses can be used to check on chain whether they own the necessary funds to fulfill the community's permissions

The checking of required funds is **not** part of this commit. It will be added in a follow-up commit.

Samyoul reviewed (approved):
Cool. Looks all good to me. The only issue is the growth of the `NewManager` func signature. But I've left a comment about in the new PR.

jrainville commented on protocol/communities/persistence.go:
```diff
@@ -322,6 +322,59 @@ func (p *Persistence) SaveRequestToJoin(request *RequestToJoin) (err error) {
+		_ = tx.Rollback()
+	}()
+
+	query := `INSERT OR REPLACE INTO communities_requests_to_join_revealed_addresses (request_id, address, signature) VALUES (?, ?, ?)`
```
I don't think saving the signature is useful. Once we have confirmed that the addresses belong to the account, I think we can just save the addresses? Maybe I forgot a use case though.

0x-r4bbit replied:
Actually that's true.
Originally I kept it to re-validate given addresses when memberships have to be accepted manually.
But I don't think there's a reason to do so anymore. revealed addresses only end up in the database if they have been successfully recovered from the signature in the first place.

Will remove that.

jrainville reviewed (commented):
Good job. I just had a question whether to save the signature or not.

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -81,7 +83,19 @@ func (t *HistoryArchiveDownloadTask) Cancel() {
+	return func(opts *managerOptions) {
+		opts.accountsManager = accountsManager
+	}
+}
```
@Samyoul FYI introduced `managerOptions` here. This will extend in the follow-up PRs
